model/src/main/kotlin/utils/DependencyGraphBuilder.kt

```diff
@@ -362,21 +362,26 @@ class DependencyGraphBuilder<D>(
             addDependencyToGraph(scopeName, it, transitive = true)
         }
 
-        val fragmentMapping = referenceMappings[index.fragment]
-        val ref = DependencyReference(
-            pkg = index.root,
-            fragment = index.fragment,
-            dependencies = transitiveDependencies.toSortedSet(),
-            linkage = dependencyHandler.linkageFor(dependency),
-            issues = issues
-        )
-        fragmentMapping[index.root] = ref
+        return referenceMappings[index.fragment].takeIf { index.root !in it }?.let { fragmentMapping ->
```

Comment:
Probably just a matter of taste, but instead of more nesting I'd prefer to return early in the `index.root in it` case, like

    val fragmentMapping = referenceMappings[index.fragment]
    if (index.root in fragmentMapping) {
        // ...
        return addDependencyToGraph(scopeName, dependency, transitive)
    }